R/taxize_nexml.R

```diff
@@ -6,38 +6,47 @@
 #' containing the corresponding identifier. 
 #' @param nexml a nexml object
 #' @param type the name of the identifier to use
-#' @param ... additional arguments (not implemented yet)
+#' @param ... additional arguments to `[taxald::get_ids()]`
 #' @export 
 #' @examples \dontrun{
 #' data(bird.orders)
 #' birds <- add_trees(bird.orders)
 #' birds <- taxize_nexml(birds, "NCBI")
 #' }
 # @importFrom taxize get_uid
-taxize_nexml <- function(nexml, type = c("NCBI"), ...){
+taxize_nexml <- function(nexml, type = c("ncbi", "itis", "col", "tpl",
+                                         "gbif", "wd"), ...){
   
   ## Soft dependency on taxize
-  if (!requireNamespace("taxize", quietly = TRUE)) {
-    stop("taxize package required to convert look up taxonomic ids",
+  if (!requireNamespace("taxald", quietly = TRUE)) {
+    stop("taxald package required to convert look up taxonomic ids",
          call. = FALSE)
   }
-  get_uid <- getExportedValue("taxize", "get_uid")
+  get_ids <- getExportedValue("taxald", "get_ids")
   
-  
-          type <- match.arg(type)
-          if(type == "NCBI"){
-            for(j in 1:length(nexml@otus)){
-              for(i in 1:length(nexml@otus[[j]]@otu)){
-                id <- get_uid(nexml@otus[[j]]@otu[[i]]@label)
-                if(is.na(id))
-                  warning(paste("ID for otu", nexml@otus[[j]]@otu[[i]]@label, "not found. Consider checking the spelling or alternate classification"))
-                else 
-                  nexml@otus[[j]]@otu[[i]]@meta <- New("ListOfmeta", list(
-                                 meta(href = paste0("http://ncbi.nlm.nih.gov/taxonomy/", id),
-                                      rel = "tc:toTaxon")))
+  type <- tolower(type)
+  type <- match.arg(type)
+
+  for(j in 1:length(nexml@otus)){
+    
+    # Resolve all ids at once
+    labels <- unname(vapply(nexml@otus[[j]]@otu, slot, character(1L), "label"))
+    taxa_ids <- get_ids(labels, type, format = "uri", ...)
+    
+    for(i in 1:length(taxa_ids)){
+      id <- taxa_ids[[i]]
+      if(is.na(id))
+        warning(paste("ID for otu", 
+                      nexml@otus[[j]]@otu[[i]]@label, 
+                      "not found. Consider checking the spelling
+                      or alternate classification"))
+      else 
+        nexml@otus[[j]]@otu[[i]]@meta <- New("ListOfmeta", list(
```

Comment:
nice catch, that's a good point.  You've been playing more with S4 more recently then me -- is there a more concise way to fix that than checking if `nexml@otus[[j]]@otu[[i]]@meta` is empty?

Reply:
I don't think you need to care whether it's empty – simply combine (`c()`) the current value (which is of type `ListOfmeta`) with the new `meta`, and the generic method for `c(ListOfmeta, ...)` should catch it and treat it correctly. No?

Reply:
```R
> length(c(new("ListOfmeta"), meta("foo:foo", "bar")))
[1] 1
> length(c(new("ListOfmeta", list(meta("baz","bang"))), meta("foo:foo", "bar")))
[1] 2
```
So that works as I claimed 😄

Reply:
nice, have now implemented this.